src/calendarevent.cpp

```diff
@@ -89,168 +99,223 @@ static Qt::TimeSpec toTimeSpec(const QDateTime &dt)
+    return mData->externalInvitation;
 }
 
 QDateTime CalendarEvent::recurrenceId() const
 {
-    return mRecurrenceId;
+    return mData->recurrenceId;
 }
 
 QString CalendarEvent::recurrenceIdString() const
 {
-    if (mRecurrenceId.isValid()) {
-        return CalendarUtils::recurrenceIdToString(mRecurrenceId);
+    if (mData->recurrenceId.isValid()) {
+        return CalendarUtils::recurrenceIdToString(mData->recurrenceId);
     } else {
         return QString();
     }
 }
 
-// Returns the event as a iCalendar string
-QString CalendarEvent::iCalendar(const QString &prodId) const
+CalendarStoredEvent::CalendarStoredEvent(CalendarManager *manager, const CalendarData::Event *data)
+    : CalendarEvent(data, manager)
+    , mManager(manager)
 {
-    if (mUniqueId.isEmpty()) {
-        qWarning() << "Event has no uid, returning empty iCalendar string."
-                   << "Save event before calling this function";
-        return QString();
-    }
+    connect(mManager, SIGNAL(notebookColorChanged(QString)),
+            this, SLOT(notebookColorChanged(QString)));
+    connect(mManager, SIGNAL(eventUidChanged(QString,QString)),
+            this, SLOT(eventUidChanged(QString,QString)));
+}
 
-    return mManager->convertEventToICalendarSync(mUniqueId, prodId);
+CalendarStoredEvent::~CalendarStoredEvent()
+{
 }
 
-void CalendarEvent::notebookColorChanged(QString notebookUid)
+void CalendarStoredEvent::notebookColorChanged(QString notebookUid)
 {
-    if (mManager->getEvent(mUniqueId, mRecurrenceId).calendarUid == notebookUid)
+    if (mData->calendarUid == notebookUid)
         emit colorChanged();
 }
 
-void CalendarEvent::eventUidChanged(QString oldUid, QString newUid)
+void CalendarStoredEvent::eventUidChanged(QString oldUid, QString newUid)
 {
-    if (mUniqueId == oldUid) {
-        mUniqueId = newUid;
+    if (mData->uniqueId == oldUid) {
+        mData->uniqueId = newUid;
         emit uniqueIdChanged();
         // Event uid changes when the event is moved between notebooks, calendar uid has changed
         emit calendarUidChanged();
     }
 }
+
+bool CalendarStoredEvent::sendResponse(int response)
+{
+    return mManager->sendResponse(*mData, (Response)response);
+}
+
+void CalendarStoredEvent::deleteEvent()
+{
+    mManager->deleteEvent(mData->uniqueId, mData->recurrenceId, QDateTime());
+    mManager->save();
+}
+
+// Returns the event as a iCalendar string
+QString CalendarStoredEvent::iCalendar(const QString &prodId) const
+{
+    Q_UNUSED(prodId);
+    if (mData->uniqueId.isEmpty()) {
+        qWarning() << "Event has no uid, returning empty iCalendar string."
+                   << "Save event before calling this function";
+        return QString();
+    }
+
+    return mManager->convertEventToICalendarSync(mData->uniqueId, prodId);
+}
+
+QString CalendarStoredEvent::color() const
+{
+    return mManager->getNotebookColor(mData->calendarUid);
+}
+
+void CalendarStoredEvent::setEvent(const CalendarData::Event *data)
```

Comment:
Hm, could this be a const reference here? Would avoid any interpretation that this would involve any instance ownership etc.

Reply:
The problem here is a cross definition : CalendarData::Event is using enums defined in CalendarEvent class.

To avoid this difficulty, I'm using a pointer. It's ugly. But transitory only. Then, when moving to KCalendarCore::Incidence::Ptr (since dropping CalendarData::Event completely), we can use references.